File: 33_recursion/33_3_powers_mod_m.cpp
#include <iostream>

int powers_mod_m(int a, int p, int m) {
  if (p == 0) {
    return 1;
  }

  int mod = a * powers_mod_m(a, p - 1, m) % m;
  return mod;
}

int main() {
  // Example1
  int a = 2;
  int p = 5;
  int m = 100;
  std::cout << powers_mod_m(a, p, m) << std::endl;

  // Example2
  a = 2;
  p = 5;
  m = 30;
  std::cout << powers_mod_m(a, p, m) << std::endl;

  return 0;
}

File: 33_recursion/33_3_powers_mod_m.py
def powers_mod_m(a: int, p: int, m: int) -> int:
  if p == 0:
    return 1
  
  mod = a * powers_mod_m(a, p-1, m) % m
  return mod

if __name__ == "__main__":
  # Example1
  a = 2
  p = 5
  m = 100
  print(powers_mod_m(a, p, m))

  # Example2
  a = 2
  p = 5
  m = 30
  print(powers_mod_m(a, p, m))